app/models/spree/stock/quantifier.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Spree
+  module Stock
+    class Quantifier
+      attr_reader :stock_items
+
+      def initialize(variant)
+        @variant = variant
+        @stock_items = fetch_stock_items
+      end
+
+      def total_on_hand
+        if Spree::Config.track_inventory_levels
+          stock_items.sum(&:count_on_hand)
+        else
+          Float::INFINITY
+        end
+      end
+
+      def backorderable?
+        stock_items.any?(&:backorderable)
+      end
+
+      def can_supply?(required)
+        total_on_hand >= required || backorderable?
+      end
+
+      private
+
+      def fetch_stock_items
+        # Don't re-fetch associated stock items from the DB if we've already eager-loaded them
+        return @variant.stock_items.to_a if @variant.stock_items.loaded?
+
+        Spree::StockItem.joins(:stock_location).
+          where(:variant_id => @variant, Spree::StockLocation.table_name => { active: true })
+      end
+    end
+  end
+end
```

Comment:
I kept the commit history clear, the original Spree version of the class is in one commit, and the changes are in the next

Reply:
I just had a big trip into the spree 2 upgrade times where this class was at the core of so many headaches (managing stock in stock items for normal variants but without stock items for overridded variants).

here are some memories:
FYI we have a few places where this quantifier is avoided since spree 2:
    - Line_item#sufficient_stock? - overridden in OFN to redirect to Variant#can_supply that uses Variant#total_on_hand
    - Stock::AvailabilityValidator#validate - overridden in OFN (for a couple of other reasons including) to redirect to Variant#can_supply? that uses Variant#total_on_hand
    - Variant#in_stock?  - overridden in OFN to redirect to Variant#can_supply? that uses Variant#total_on_hand
    - **Variant#total_on_hand - this is the only place where Quantifier is used** and only for not-overrided variants (variant override does a different thing without StockItems).

Reply:
do we have that know-how written down somewhere @luisramos0 ? If I had to find out all that it'd took quite a lot of time :see_no_evil:

Reply:
this is an analysis of all calls to Quantifer. It's basically saying that during spree upgrade most methods that used this Quantifier class were overriden in OFN.
Bye bye spree will just remove all this complexity because there will only be one call to Quantifier :+1: